rednose/helpers/ekf_sym.cc

```diff
@@ -205,3 +205,180 @@ VectorXd EKFSym::update(int kind, VectorXd z, MatrixXdr R, std::vector<double> e
   }
   return z;
 }
+
+void EKFSym::augment() {
+    // TODO this is not a generalized way of doing this and implies that the
+    // augmented states are simply the first (dim_augment_state) elements of
+    // the main state.
+
+    // TODO remove commented python code after testing pass
+    assert(this->msckf);
+
+    int d1 = this->dim_main;
+    int d2 = this->dim_main_err;
+    int d3 = this->dim_augment;
+    int d4 = this->dim_augment_err;
+
+    // push through augmented states
+    // this->x[d1:-d3] = this->x[d1 + d3:];
+    // ^ make elements from d1 to (element d3 elements far from the end)
+    // Traversal must be in reverse to prevent overwriting data
+    // Start at end-d3, work way back to d1
+    int x_size = this->x.size();
+    for (int ix = (x_size-1)-d3; ix >= d1; ix--) {
+        this->x[ix] = this->x[ix+d3];
+    }
+
+    // this->x[-d3:] = this->x[:d3];
+    // ^ last d3 elements = first d3 elements
+    for (int ix = 0; ix < d3; ix++) {
+        this->x[x_size-d3+ix] = this->x[ix];
+    }
+
+    assert(this->x.rows() == this->dim_x);
+    assert(this->x.cols() == 1);
+
+    // push through augmented covs
+    assert(this->P.rows() == this->dim_err);
+    assert(this->P.cols() == this->dim_err);
+
+    // Build P_reduced by selectively copying P's contents
+    // Skip rows from d2 to before (d2+d4), do same with columns
+    int red_side = this->P.rows() - d4;
+    MatrixXd P_reduced = MatrixXd::Zero(red_side, red_side);
+    MatrixXd to_mult   = MatrixXd::Zero(this->dim_err, red_side);
+    int irow,        icol;
+    int irow_offset, icol_offset;
+    for (int i = 0; i < red_side; i++) {
+        // Build P_reduced
+        icol = i % this->dim_err;
+        irow = i / this->dim_err;
+
+        icol_offset = (icol >= d2) ? d4 : 0;
+        irow_offset = (irow >= d2) ? d4 : 0;
+
+        P_reduced(irow,icol) = P(irow+irow_offset,icol+icol_offset);
+
+        // Build to_mult
+        // to_mult[:-d4, :] = np.eye(self.dim_err - d4)
+        // The submatrix comprised of all rows except last d4 is set to an identity
+        // matrix
+        to_mult(i, i) = 1;
+        // to_mult[-d4:, :d4] = np.eye(d4)
+        // The submatrix comprised by the last d4 rows and therein all columns up
+        // to the d4-th column is set to an identity matrix
+        if (i < d4) to_mult(red_side+i, i) = 1;
+    }
+
+    // TODO aren't we certain enough that the side of P_reduced is red_side?
+    // Are these asserts really required?
+    assert(P_reduced.rows() == red_side);
+    assert(P_reduced.cols() == red_side);
```

Comment:
@pd0wm What do you think about removing these asserts, inherited from the python code? Matrix dimensions defined in line 248.